### PR TITLE
Allow reading complex numbers as hdf5 compound datatype + test

### PR DESCRIPTION
**Pull request by Wentzell**

### CODE REVIEW

File: c++/h5/object.cpp [hunk 1]
```diff
@@ -43,17 +43,28 @@ namespace h5 {
...
-  template <> hid_t hdf5_type<const char *> (){return  details::str_dtype;}
+  template <> hid_t hdf5_type<std::string>  (){return  details::str_dt;}
+  template <> hid_t hdf5_type<char *>       (){return  details::str_dt;}
+  template <> hid_t hdf5_type<const char *> (){return  details::str_dt;}
+
+  namespace details {
+    hid_t const cplx_cmpd_dt = [](){
+      hid_t dt = H5Tcreate(H5T_COMPOUND, 16);
+      H5Tinsert(dt, "r", 0, H5T_IEEE_F64LE);
+      H5Tinsert(dt, "i", 8, H5T_IEEE_F64LE);
```
Comment (parcollet) [inline c++/h5/object.cpp]: why not H5T_NATIVE_DOUBLE ??

File: c++/h5/object.hpp [hunk 2]
```diff
@@ -18,6 +18,12 @@ namespace h5 {
   template <typename T>
   hid_t hdf5_type();
 
+  // A complex compound type consisting of two doubles
+  // This type is stored and loaded as an hdf5 compound datatype
+  struct cplx_t {
```
Comment (parcollet) [inline c++/h5/object.hpp]: --> dcplx_t ? 
there are double ....

File: c++/h5/scalar.hpp [hunk 3]
```diff
@@ -30,7 +31,18 @@ namespace h5 {
...
       }
     }
-    array_interface::read(g, name, array_interface::h5_array_view_from_scalar(x), array_interface::get_h5_lengths_type(g, name));
+
+    auto lt = array_interface::get_h5_lengths_type(g, name);
+
+    if constexpr (is_complex_v<T>) {
+      // Allow reading complex as a compound hdf5 dataype
+      if (hdf5_type_equal(lt.ty, hdf5_type<cplx_t>())) {
+        h5_read(g, name, reinterpret_cast<cplx_t &>(x));
```
Comment (parcollet) [inline c++/h5/scalar.hpp]: reinterpret_cast : most likely correct for complex, 
you can also simply copy to be very safe, the optimizer will remove the copy anyway

File: test/c++/CMakeLists.txt [hunk 4]
```diff
@@ -11,7 +11,7 @@ foreach(test ${all_tests})
   get_filename_component(test_name ${test} NAME_WE)
   get_filename_component(test_dir ${test} DIRECTORY)
   add_executable(${test_name} ${test})
-  target_link_libraries(${test_name} ${PROJECT_NAME}_c gtest_main ${PROJECT_NAME}_warnings)
+  target_link_libraries(${test_name} ${PROJECT_NAME}_c h5::hdf5 gtest_main ${PROJECT_NAME}_warnings)
```
Comment (parcollet) [inline test/c++/CMakeLists.txt:14]: why this ?

File: c++/h5/stl/array.hpp [hunk 5]
```diff
@@ -60,27 +60,31 @@ namespace h5 {
...
+          h5_read(g, name, reinterpret_cast<std::array<cplx_t, N> &>(a));
+          return;
+        }
+
+        // Allow to read non-complex data into array<complex>
         if (!lt.has_complex_attribute) {
           std::array<double, N> tmp;
           h5_read(g, name, tmp);
-          a = tmp;
+          std::copy(begin(tmp), end(tmp), begin(a));
```
Comment (parcollet) [inline c++/h5/stl/array.hpp:82]: why this change ?
a = b is fine for a std::array, and simpler ...

File: test/c++/CMakeLists.txt (hunk 4 above)
Comment (Wentzell) [reply test/c++/CMakeLists.txt:14]: just for the tests

File: c++/h5/object.cpp (hunk 1 above)
Comment (Wentzell) [reply c++/h5/object.cpp]: yes changed
